### PR TITLE
Adjust to ixmp Backend/Model refactor

### LINKED ISSUE #238: Set up Codecov CI configuration and badge
This can be done with reference to iiasa/ixmp#192.

### PR DESCRIPTION
**Pull request by khaeru**
iiasa/ixmp#182 modularizes `ixmp.core` into abstract Backend and Model classes, and concrete JDBCBackend and GAMSModel classes. This PR adjusts message_ix accordingly.

## How to review
- Read the docs preview: https://message-ix-khaeru.readthedocs.io/en/feature-ixmp-backend/
- Run the tests.
- Look at the code/diff.
  - The CI-related files (.appveyor.yml, .travis.yml, ci/*, setup.cfg and setup.py) can be skimmed/ignored; they mainly copy what is done in ixmp.

## Details
message_ix jumps stack layers in a few places by communicating directly with ixmp_source.
<details><summary>Click to show source…</summary>
- MsgScenario.getTypeList in `cat_list`: https://github.com/iiasa/message_ix/blob/570f79833eddf046f7779db3f68fb86c0b13b0cc/message_ix/core.py#L94
- MsgScenario.addCatEle in `add_cat`: https://github.com/iiasa/message_ix/blob/570f79833eddf046f7779db3f68fb86c0b13b0cc/message_ix/core.py#L111
- MsgScenario.getCatEle in `cat`: https://github.com/iiasa/message_ix/blob/570f79833eddf046f7779db3f68fb86c0b13b0cc/message_ix/core.py#L123
- MsgScenario.getFirstModelYear in `firstmodelyear`: https://github.com/iiasa/message_ix/blob/570f79833eddf046f7779db3f68fb86c0b13b0cc/message_ix/core.py#L272
- MsgScenario.getTecActYrs. Note this was incorrectly in **ixmp**.Scenario; iiasa/ixmp#182 deletes it in ixmp.
</details>

The code here:
- Modifies message_ix.Scenario:
  - Adds a method `._backend(name, ...)` that calls backend API methods named `ixmp.Backend.ms_{name}(...)`.
  - Moves `.years_active()` from ixmp.Scenario.
    - Instead of calling the ixmp_source method getTecActYrs (closed source; perpetuates a stack level violation in the Backend), I've just reimplemented this (open-source) in message_ix.
  - Deletes `.remove_solution()`—it only duplicated ixmp.Scenario.
  - Significantly simplifies .clone(). Now:
    - message_ix.Scenario.clone only exists as a placeholder for its docstring.
    - ixmp.Scenario.clone handles valid combinations of arguments; so these are the same for ixmp.Scenario and message_ix.Scenario
    - JDBCBackend.s_clone raises exceptions for specific limitations of ixmp_source: https://github.com/iiasa/ixmp/blob/962f739bf71a1718b1c3fdf8f6e73e983986fc6e/ixmp/backend/jdbc.py#L321-L333
- Deletes `message_ix.model_settings` and adds `message_ix.models` containing two subclasses of `ixmp.GAMSModel` named `MESSAGE` and `MESSAGE_MACRO`.
- Also:
  - message_ix.default_model_paths is simplified using `pathlib`. This should eventually be delegated to ixmp.config.Config.
  - Handling of CPLEX optfiles is moved into the new `MESSAGE` model class.
  - ixmp documentation is not repeated in message_ix; instead, linked with intersphinx.
  - CI configuration is revised to mirror ixmp; this includes adding Codecov, thus closes #238.

## PR checklist

- [x] Tests added.
- [x] Documentation added.
- [x] Release notes updated.

### CODE REVIEW

File: doc/source/api.rst [hunk 1]
```diff
@@ -1,17 +1,111 @@
...
+   These configure the GAMS CPLEX solver (or another solver, if selected); see `the solver documentation <https://www.gams.com/latest/docs/S_CPLEX.html>`_ for possible values.
+
+.. autoclass:: MESSAGE
+   :members:
+   :show-inheritance:
+
+   The MESSAGE Python class encapsulates the GAMS code for the core MESSAGE mathematical formulation.
+   The *model_options* arguments are received from :meth:`.Scenario.solve`, and—except for *solve_options*—are passed on to the parent class :class:`~ixmp.model.gams.GAMSModel`; see there for a full list of options.
+
+.. autoclass:: MESSAGE_MACRO
```
Comment (gidden) [inline doc/source/api.rst]: This looks to be broken on RTD? https://message-ix-khaeru.readthedocs.io/en/feature-ixmp-backend/api.html#message_ix.models.MESSAGE.defaults

Comment (khaeru) [reply doc/source/api.rst]: Good eye. The code is working as intended, but that's indeed not what we want the user to be seeing in the documentation. I'll try for a fix.

**Comment by khaeru** [comment]
> how will this play with the MACRO PR #223?

A rebase/merge from `master` might be needed after this goes in. I'll add some comments over there.

**Comment by khaeru** [comment]
See the above comment on IAMconsortium/pyam#281; IAMconsortium/pyam#280 broke the build, so I will now try to add a workaround.

**Comment by khaeru** [comment]
Okay, that worked. Thanks @gidden for the quick review.

**Pull request merged**